### PR TITLE
Add support for Muse Glimmer and its tool format

### PR DESCRIPTION
**Pull request by neilmehta24**
- Update mlx-vlm pin to add support for Muse Glimmer
- Add support for glimmer's tool calling grammar, in order to constrain invalid logits during tool call generation
- Update the batch generator to eval the prefill arrays before moving on to decode, in order to fix the prefill/decode latency attribution. Previously, the final prefill work was being tracked as the first decode token latency, which misattributes the latency profiles. E2E request completion time is unaffected.

**Performance on M2 Mac Studio:**
- Decode: **31.5 tok/s** at low context; 29 tok/s at 8k context length
- Prefill: **175 tok/s** for 8k context

### CODE REVIEW

File: mlx_engine/tool_runtime.py [hunk 1]
```diff
@@ -234,21 +298,55 @@ def process_last_token_with_context(
...
+    def _tool_call_start_condition(
+        self, token_context: list[int], token_id: mx.array
+    ) -> mx.array:
+        marker_prefix = self._tool_call_start_token_ids[:-1]
+        if len(marker_prefix) > len(token_context):
+            return mx.array(False)
+        if (
+            len(marker_prefix) > 0
+            and tuple(token_context[-len(marker_prefix) :]) != marker_prefix
+        ):
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/tool_runtime.py:347]: > That lets prompt content trigger a valid tool call instead of only constraining tool calls wholly generated by the assistant

This is fine. And very rare

File: mlx_engine/tool_runtime.py [hunk 2]
```diff
@@ -98,6 +106,54 @@ def create_qwen35_tool_context_from_prompt(
...
 
 
+def create_muse_glimmer_tool_context_from_prompt(
+    *,
+    tokenizer: Any,
+    prompt_tokens: list[int],
+    model_type: str | None,
+) -> MuseGlimmerToolContext | None:
+    """Return Muse Glimmer context for a prompt with native ATEM tools."""
+    if model_type != "muse_glimmer":
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/tool_runtime.py:116]: This is acceptable, there aren't any text-only conversions floating around

**Review by will-lms — APPROVED** [review]
